typing/types.mli

```diff
@@ -408,6 +369,45 @@ end
 
 val create_expr: type_desc -> level: int -> scope: int -> id: int -> type_expr
 
+(** This module provides the interface to construct, query, and destruct
+    [with_bounds_types] and [nonempty_with_bounds_types]. Under the hood this is a
+    [Stdlib.Map] from [type_expr] to [With_bounds_type_info.t], using a "best-effort"
+    semantic comparison on [type_expr] to provide the mapping. This "best-effort"ness
+    means that two semantically equal (according to [Ctype.eqtype]) types might have
+    distinct keys in the map - but two semantically {i inequal} types are guaranteed never
+    to occupy the same key.
+*)
+module With_bounds_types : sig
+  type info := With_bounds_type_info.t
+  type t := with_bounds_types
+
+  val empty : t
+  val is_empty : t -> bool
+  val to_seq : t -> (transient_expr * info) Seq.t
```

Comment:
But `transient_expr` is meant to be transient. More precisely, a `transient_expr` becomes invalid once any mutation happens. `with_bounds` exist for a long time, and so they shouldn't hold `transient_expr`s.

Instead, I think using `type_expr` here was right, but we just need to `repr` more aggressively. This is probably best done by using `get_desc`, as `eqtype` does.

Reply:
ah, yeah, good point. Updated to just call `get_desc` in the comparator.